kiwi/api_helper.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) 2021 SUSE Software Solutions Germany GmbH. All rights reserved.
+#
+# This file is part of kiwi.
+#
+# kiwi is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# kiwi is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with kiwi.  If not, see <http://www.gnu.org/licenses/>
+#
+import logging
+from functools import wraps
+from typing import Callable
+
+from kiwi.exceptions import KiwiDecommissionedApiFunctionCalled
+
+log = logging.getLogger('kiwi')
+
+
+def obsolete(decommission_at: str) -> Callable:
+    """
+    Decorator method to mark an API method as obsolete
+
+    Methods marked with this decorator are still called.
+    If the provided decommission date is reached it is
+    allowed to decorate the obsoleted method as
+    decommissioned.
+
+    Example:
+
+        .. code:: python
+
+            @obsolete(decommission_at='2025-01-28')
+            def method():
+                print('Method implementation...')
+
+    :param str decommission_at: date string
+    """
+    def obsolete_method(func: Callable) -> Callable:
```

Comment:
I think the `method` in the function name could be a bit misleading and overloaded. :wink:  When I write decorators, I usually follow this naming scheme:

```python
def mydecorator(an_argument):
    """ ... """
    def decorate_mydecorator(func):
         @wraps(func)
         def wrapper(*args, **kwargs):
             # ...
             return func(*args, **kwargs)
         return wrapper
     return decorate_mydecorator
```

Applying this principle to your code, I would suggest:

```suggestion
    def decorate_obsolete(func: Callable) -> Callable:
```

Reply:
I have seen this style at many places but was not so happy with the name "wrapper". It's actually a wrapper but the decorator itself is also a wrapper. And if you are using decorators with params you have the wrapper of the wrapper. But as I see this style at several places I will use it too. It seems to be kind of a common standard. Thanks will update the code

Reply:
You are basically right, I had the same sentiments. I explained it to myself that `wrapper` is the inner wrapper that is really applied to the decorated function. The others are just, well, some supportive "wrappers". :smile:

Reply:
done now